File: kube-devops-backend/fastApiProject/routers/overview.py
# routers/overview.py
from fastapi import APIRouter
from datetime import datetime, timezone

from services.kube_client import get_cluster_counts
from services.alert_client import list_alerts
from services.prometheus_client import instant_value

router = APIRouter(prefix="/api", tags=["overview"])


def r1(v: float | None) -> float:
    """统一保留 1 位小数，防 Prometheus 抖动 / None"""
    return round(v or 0.0, 1)


@router.get("/overview")
def get_overview():
    # ===============================
    # 1. Kubernetes 资源统计
    # ===============================
    counts = get_cluster_counts()

    # ===============================
    # 2. 告警统计（Alertmanager）
    # ===============================
    try:
        alerts = list_alerts() or []
    except Exception:
        alerts = []

    firing = sum(
        1 for a in alerts
        if (a.get("status") or {}).get("state") == "active"
    )

    # ===============================
    # 3. Prometheus 使用率（集群级）
    # ===============================

    # CPU 使用率（%）
    cpu_usage = instant_value("""
    100 * (1 - avg(rate(node_cpu_seconds_total{mode="idle"}[5m])))
    """)

    # 内存使用率（%）—— 容器视角 / 节点总内存
    memory_usage = instant_value("""
    100 *
    sum(container_memory_working_set_bytes{container!="",pod!=""})
    /
    sum(node_memory_MemTotal_bytes)
    """)

    # 存储使用率（%）—— 宿主机磁盘（过滤 pseudo FS）
    storage_usage = instant_value("""
    100 *
    (
      1 -
      sum(node_filesystem_avail_bytes{
        fstype!~"tmpfs|overlay|squashfs|nsfs"
      })
      /
      sum(node_filesystem_size_bytes{
        fstype!~"tmpfs|overlay|squashfs|nsfs"
      })
    )
    """)

    # 网络吞吐（MB/s）—— 非百分比
    network_throughput = instant_value("""
    (
      sum(rate(node_network_receive_bytes_total{device!="lo"}[5m])) +
      sum(rate(node_network_transmit_bytes_total{device!="lo"}[5m]))
    ) / 1024 / 1024
    """)

    # 数值规整
    cpu_usage = r1(cpu_usage)
    memory_usage = r1(memory_usage)
    storage_usage = r1(storage_usage)
    network_throughput = r1(network_throughput)

    # ===============================
    # 4. 返回 BFF 聚合视图模型
    # ===============================
    return {
        "overviewStats": [
            {"key": "nodes", "label": "节点", "value": counts["nodes"]},
            {"key": "namespaces", "label": "命名空间", "value": counts["namespaces"]},
            {"key": "pods", "label": "Pod", "value": counts["pods"]},
            {"key": "services", "label": "Service", "value": counts["services"]},
            {"key": "workloads", "label": "工作负载", "value": counts["workloads"]},
            {
                "key": "alerts24h",
                "label": "告警",
                "value": firing,
                "desc": "当前活跃告警",
            },
        ],
        "capacityStats": [
            {
                "key": "cpu",
                "label": "CPU 使用率",
                "value": cpu_usage,
                "unit": "%",
                "desc": "集群 CPU 整体忙碌程度",
            },
            {
                "key": "memory",
                "label": "内存使用率",
                "value": memory_usage,
                "unit": "%",
                "desc": "容器实际内存使用占节点总内存比例",
            },
            {
                "key": "storage",
                "label": "存储使用率",
                "value": storage_usage,
                "unit": "%",
                "desc": "宿主机磁盘空间使用情况",
            },
            {
                "key": "network",
                "label": "网络吞吐",
                "value": network_throughput,
                "unit": "MB/s",
                "desc": "集群网络实时吞吐速率",
            },
        ],
        "opsSummary": [
            {
                "title": f"当前有 {firing} 条告警",
                "desc": "来自 Alertmanager",
                "level": "warn" if firing > 0 else "ok",
            }
        ],
        "cluster": {
            "basic": {
                "name": "host",
                "status": "running",
                "clusterType": "K3s",
                "k8sVersion": "v1.28.x",
                "platformVersion": "v1.0.0-rc.1",
                "location": "本地实验环境",
            },
            "resources": {
                "nodes": counts["nodes"],
                "namespaces": counts["namespaces"],
                "pods": counts["pods"],
                "alertRules": firing,
            },
            "health": [
                {"level": "ok", "text": "API Server 正常"},
                {"level": "ok", "text": "Scheduler 正常"},
            ],
        },
        "updatedAt": datetime.now(timezone.utc).isoformat(),
    }

File: kube-guard-frontend/src/api/overview.ts
// src/api/overview.ts
import { get } from './http'

export type OverviewStatItem = {
  key: string
  label: string
  value: number
  unit?: string
  desc?: string
}

export type CapacityItem = {
  key: 'cpu' | 'memory' | 'storage' | 'network'
  label: string
  value: number // 0-100
  unit: '%'
  desc?: string
}

export type OpsItem = {
  title: string
  desc?: string
  extra?: string
  level: 'ok' | 'warn' | 'info'
}

export type ClusterBasic = {
  name: string
  status: 'running' | 'warning' | 'down'
  clusterType?: string
  k8sVersion?: string
  platformVersion?: string
  location?: string
}

export type ClusterResourceSummary = {
  nodes: number
  namespaces: number
  pods: number
  alertRules?: number
}

export type ClusterHealthItem = {
  level: 'ok' | 'warn' | 'error'
  text: string
}

export type OverviewResponse = {
  overviewStats: OverviewStatItem[]
  capacityStats: CapacityItem[]
  opsSummary: OpsItem[]
  cluster: {
    basic: ClusterBasic
    resources: ClusterResourceSummary
    health: ClusterHealthItem[]
  }
}

/**
 * ✅ 建议后端提供一个聚合接口：
 * GET /api/overview
 *
 * 后端可以内部去 Prometheus / Alertmanager / K8s API 聚合出：
 * - 概览数字（nodes/pods/workloads）
 * - CPU/内存/存储/网络 使用率（Prometheus）
 * - 最近告警摘要（Alertmanager）
 * - 集群健康点检（K8s API）
 */
export function fetchOverview() {
  return get<OverviewResponse>('/api/overview')
}
